gateway-service-impl/src/main/java/org/hypertrace/gateway/service/baseline/BaselineServiceImpl.java

```diff
@@ -0,0 +1,383 @@
+    List<TimeAggregation> timeAggregations = new ArrayList<>();
+    for (BaselineTimeAggregation timeAggregation : timeSeriesList) {
+      Expression aggregation =
+          Expression.newBuilder().setFunction(timeAggregation.getAggregation()).build();
+      TimeAggregation agg =
+          TimeAggregation.newBuilder()
+              .setAggregation(aggregation)
+              .setPeriod(timeAggregation.getPeriod())
+              .build();
+      timeAggregations.add(agg);
+    }
+    return timeAggregations;
+  }
+
+  private List<TimeAggregation> getTimeAggregationsForAggregateExpr(
+      BaselineEntitiesRequest originalRequest) {
+    List<FunctionExpression> aggregateList = originalRequest.getBaselineAggregateRequestList();
+    List<TimeAggregation> timeAggregationList = new ArrayList<>();
+    for (FunctionExpression function : aggregateList) {
+      TimeAggregation timeAggregation =
+          getAggregationFunction(
+              function, originalRequest.getStartTimeMillis(), originalRequest.getEndTimeMillis());
+      timeAggregationList.add(timeAggregation);
+    }
+    return timeAggregationList;
+  }
+
+  private TimeAggregation getAggregationFunction(
+      FunctionExpression function, long startTimeMillis, long endTimeMillis) {
+    Period period = getPeriod(startTimeMillis, endTimeMillis);
+    TimeAggregation timeAggregation =
+        TimeAggregation.newBuilder()
+            .setPeriod(period)
+            .setAggregation(Expression.newBuilder().setFunction(function).build())
+            .build();
+    return timeAggregation;
+  }
+
+  private Period getPeriod(long startTimeInMillis, long endTimeMillis) {
+    Period period;
+    if ((endTimeMillis - startTimeInMillis) < DAY_IN_MILLIS) {
+      long timePeriod = (endTimeMillis - startTimeInMillis) / (1000 * 60);
+      period = Period.newBuilder().setUnit("MINUTES").setValue((int) timePeriod).build();
+    } else {
+      long timePeriod = (endTimeMillis - startTimeInMillis) / (1000 * 60 * 60 * 24);
+      period = Period.newBuilder().setUnit("HOURS").setValue((int) timePeriod).build();
+    }
+    return period;
+  }
+
+  private Period getTimeSeriesPeriod(List<BaselineTimeAggregation> baselineTimeSeriesList) {
+    return baselineTimeSeriesList.get(0).getPeriod();
+  }
+
+  private long getPeriodInSecs(Period period) {
+    ChronoUnit unit = ChronoUnit.valueOf(period.getUnit());
+    long periodSecs = Duration.of(period.getValue(), unit).getSeconds();
+    return periodSecs;
+  }
+
+  /**
+   * If user selected time range is less than 24 hrs then choose last 24hrs of data for baseline if
+   * it is gt 24hours then choose 2X time range for baseline calculation.
+   *
+   * @param startTimeMillis
+   * @param endTimeMillis
+   * @return
+   */
+  private long getUpdatedStartTimeForAggregations(long startTimeMillis, long endTimeMillis) {
+    if ((endTimeMillis - startTimeMillis) < DAY_IN_MILLIS) {
+      startTimeMillis = endTimeMillis - DAY_IN_MILLIS;
```

Comment:
shift should be based on whole range (i.e. for TR [t_0 -> t_1], we want baseline data of [t_0-24h, t_0]. This implementation gives us [t_1-24h, t_1] for baseline data.

Reply:
I think this one is important to fix before merging (see my comment above)

Reply:
Since we used now Max(24h, 2XtimeRange) even we are using [t_1-24h, t_1] it will give decent data for baseline and also if we don't include current data we might increase in traffic scenarios. We can also check how this is going and iterate on this.

Reply:
Responded to this one on one of the other threads - https://github.com/hypertrace/gateway-service/pull/55#discussion_r552241794

Reply:
Fixed this.